keeper/keeper.go

```diff
@@ -54,26 +62,80 @@ func NewKeeper(
 		ClientID:                collections.NewItem(sb, availblob1.ClientIDKey, "client_id", collections.StringValue),
-		ProvenHeight:            collections.NewItem(sb, availblob1.ProvenHeightKey, "proven_height", collections.Int64Value),
+		ProvenHeight:            collections.NewItem(sb, availblob1.ProvenHeightKey, "proven_height", collections.Uint64Value),
 		PendingBlocksToTimeouts: collections.NewMap(sb, availblob1.PendingBlocksToTimeouts, "pending_blocks_to_timeouts", collections.Int64Key, collections.Int64Value),
 		TimeoutsToPendingBlocks: collections.NewMap(sb, availblob1.TimeoutsToPendingBlocks, "timeouts_to_pending_blocks", collections.Int64Key, codec.CollValue[types.PendingBlocks](cdc)),
 
 		storeKey: key,
 
 		cdc: cdc,
 
-		publishToAvailBlockInterval: publishToAvailBlockInterval,
-		app_id:                      appId,
+		appID: appId,
 
-		unprovenBlocks: make(map[int64][]byte),
+		unprovenBlocks:              make(map[int64][]byte),
+		MaxBlocksForBlob:            20, //Todo: call this from app.go, later change to params
+		PublishToAvailBlockInterval: 5,  //Todo: call this from app.go, later change to params
+		VotingInterval:              5,
 	}
 }
 
 func (k *Keeper) SetRelayer(r *relayer.Relayer) {
 	k.relayer = r
 }
+
+func (k *Keeper) GetBlobStatus(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+	return GetStatusFromStore(store)
+}
+
+func (k *Keeper) UpdateBlobStatus(ctx sdk.Context, req *types.MsgUpdateBlobStatusRequest) (*types.MsgUpdateBlobStatusResponse, error) {
+	// status should be changed to Voting or Ready, depending on the request
+	store := ctx.KVStore(k.storeKey)
+	provenHeight := k.GetProvenHeightFromStore(ctx)
+	endHeight := k.GetEndHeightFromStore(ctx)
+	status := GetStatusFromStore(store)
+
+	if req.BlocksRange.From != provenHeight+1 || req.BlocksRange.To != endHeight {
+		return nil, fmt.Errorf("invalid blocks range request: expected range [%d -> %d], got [%d -> %d]",
+			provenHeight+1, endHeight, req.BlocksRange.From, req.BlocksRange.To)
+	}
+
+	if status != PENDING_STATE {
+		return nil, errors.New("can update the status if it is not pending")
+	}
+
+	newStatus := IN_VOTING_STATE
+	if !req.IsSuccess {
+		newStatus = FAILURE_STATE
+	} else {
+		currentHeight := ctx.BlockHeight()
+		UpdateAvailHeight(ctx, store, req.AvailHeight)                            // updates avail height at which the blocks got submitted to DA
+		UpdateVotingEndHeight(ctx, store, uint64(currentHeight)+k.VotingInterval) // TODO: Now voting interval is 5, so check whether we can process votes at next block after tx exec.
+	}
+
+	UpdateBlobStatus(ctx, store, newStatus) // updates blob status after based on tx exec
+
+	return &types.MsgUpdateBlobStatusResponse{}, nil
+}
+
+func (k *Keeper) SubmitBlobStatus(ctx sdk.Context, _ *types.QuerySubmitBlobStatusRequest) (*types.QuerySubmitBlobStatusResponse, error) {
```

Comment:
I think we can move this methods to grpc_query server and msg_server files respectively